stdlib/asm/mem.masm

```diff
@@ -0,0 +1,41 @@
+# ===== MEMORY FUNCTIONS ==========================================================================
+
+#! memcopy(read_ptr: address, write_ptr: address, words: felt)
+#!
+#! Copies `words` words from `read_ptr` to `write_ptr`.
+#!
+#! Stack transition looks as follows:
+#! [read_ptr, write_ptr, words, ...] -> [...]
+export.memcopy
+  # add counter
+  # stack: [read_ptr, write_ptr, count, words, ...]
+  push.0 movdn.2
+
+  # Pad the stack because mem_load overwrites it
+  # stack: [PAD, read_ptr, write_ptr, count, words, ...]
+  padw
+
+  # LOOP: [PAD, read_ptr, write_ptr, count, words, ...]
+  # while(words>0) (17 cycles)
+  dup.7 dup.7 neq while.true
+    # perform read (2 cycles)
+    # stack: [A, read_ptr, write_ptr, count, words, ...]
+    dup.4 mem_loadw
+
+    # peform write (3 cycles)
+    # stack: [A, read_ptr, write_ptr, count, words, ...]
+    dup.5 mem_storew
+
+    # update counters (9 cycles)
+    # stack: [read_ptr+1, write_ptr+1, count+1, words, PAD, ...]
+    swapw
+        add.1 movup.3 movup.3 add.1 movup.3 add.1 movup.3
+    swapw 
```

Comment:
I wonder if updating pointers right before we use them would be any more efficient. For example, we could do something like this:
```
while.true
  movup.4
  dup
  add.1
  movdn.5
  mem_loadw
  movup.5
  dup
  add.1
  movdn.6
  mem_storew
  ...
```
And by the time we execute `mem_storew` the pointers are already updated.

Reply:
Same thing:

```
  # LOOP: [PAD, read_ptr, write_ptr, count, n, ...]
  # while(n>0) (17 cycles)
  dup.7 dup.7 neq while.true
    # copy read_ptr and update it (4 cycles)
    # stack: [read_ptr, A, read_ptr+1, write_ptr, count, n, ...]
    movup.4 dup add.1 movdn.5

    # peform read (1 cycle)
    # stack: [A, read_ptr+1, write_ptr, count, n, ...]
    mem_loadw

    # peform write (4 cycles)
    # stack: [write_ptr, A, read_ptr+1, write_ptr+1, count, n, ...]
    movup.5 dup add.1 movdn.6

    # peform write (1 cycles)
    # stack: [A, read_ptr+1, write_ptr+1, count, n, ...]
    mem_storew

    # update counter (4 cycles)
    # stack: [A, read_ptr+1, write_ptr+1, count, n, ...]
    movup.6 add.1 dup movdn.7

    dup.8 neq # while(n>0) (3 cycles)
  end
```